### ISSUE TITLE
Incorrect construction of query with innerJoin and where (incorrect order of statement parameter)

### ISSUE THREAD

**Issue by TimWolla** [typeorm/typeorm#4429]
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[ ] `latest`
[ ] `@next`
[x] `0.2.18` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

```ts
import "reflect-metadata";
import {createConnection, MoreThan, Entity, PrimaryGeneratedColumn, ManyToMany, JoinTable, Column} from "typeorm";

@Entity()
class Bar {
    @PrimaryGeneratedColumn()
    id: number;
    @ManyToMany(() => Foo, foo => foo.bars)
    @JoinTable()
    foos: Foo[];
}

@Entity()
class Foo {
    @PrimaryGeneratedColumn()
    id: number;
    @ManyToMany(() => Bar, bar => bar.foos)
    bars: Bar[];
    @Column()
    date: Date;
}

createConnection({
    type: 'mysql',
    database: 'test',
    logging: true,
    entities: [ Foo, Bar ],
    host: '172.17.0.2',
    username: 'root',
    password: 'root'
}).then(async connection => {
    connection.getRepository(Foo)
    .createQueryBuilder()
    .where({ date: MoreThan(new Date()) })
    .innerJoin('Foo.bars', 'bar', 'bar.id = :id', { id: 1 })
    .getMany();
}).catch(error => console.log(error));
```

TypeORM constructs an incorrect query here: The placeholders are filled in the wrong order. The `id: 1` is used for the `date`, the `new Date()` is used for the ID:

```
query: SELECT `Foo`.`id` AS `Foo_id`, `Foo`.`date` AS `Foo_date` FROM `foo` `Foo` INNER JOIN `bar_foos_foo` `bar_Foo` ON `bar_Foo`.`fooId`=`Foo`.`id` INNER JOIN `bar` `bar` ON `bar`.`id`=`bar_Foo`.`barId` AND (`bar`.`id` = ?) WHERE `Foo`.`date` > ? -- PARAMETERS: ["2019-07-13T15:18:39.036Z",1]
```

And in MySQL's general log:

```
2019-07-13T15:18:39.118350Z	   48 Query	SELECT `Foo`.`id` AS `Foo_id`, `Foo`.`date` AS `Foo_date` FROM `foo` `Foo` INNER JOIN `bar_foos_foo` `bar_Foo` ON `bar_Foo`.`fooId`=`Foo`.`id` INNER JOIN `bar` `bar` ON `bar`.`id`=`bar_Foo`.`barId` AND (`bar`.`id` = '2019-07-13 15:18:39.036') WHERE `Foo`.`date` > 1
```

**Comment by TimWolla** [comment]
Digging through the code this is caused by a mixture of different placeholders:

- The join uses the named placeholder `:id`.
- The where uses question mark placeholders: `?`.
- In `MysqlDriver#escapeQueryWithParameters` the `:id` placeholder is converted into a question mark and the value is **appended** to the parameter list, without taking into account the order of existing question marks.

**Comment by danilaplee** [comment]
is also on postgresql 
```
'SELECT DISTINCT "distinctAlias"."r_id" as "ids_r_id", "distinctAlias".r_sort FROM (SELECT "r"."id" AS "r_id", "r"."englishName" AS "r_englishName", "r"."russianName" AS "r_russianName", "r"."updatedAt" AS "r_updatedAt", "r"."date" AS "r_date", "r"."type" AS "r_type", "r"."committee" AS "r_committee", "d"."id" AS "d_id", "d"."name" AS "d_name", "d"."url" AS "d_url", "d"."previewUrl" AS "d_previewUrl", "d"."sourceFileId" AS "d_sourceFileId", "d"."type" AS "d_type", "d"."previousFileId" AS "d_previousFileId", "d"."version" AS "d_version", "d"."orderId" AS "d_orderId", "d"."converted" AS "d_converted", "d"."converter_error" AS "d_converter_error", "d"."uploadedDateTime" AS "d_uploadedDateTime", "d"."path" AS "d_path", "d"."meetingId" AS "d_meetingId", "d"."reportId" AS "d_reportId", "d"."responsibleId" AS "d_responsibleId", "d"."folderId" AS "d_folderId", "u"."id" AS "u_id", "u"."englishName" AS "u_englishName", "u"."englishPosition" AS "u_englishPosition", "u"."russianName" AS "u_russianName", "u"."russianPosition" AS "u_russianPosition", "u"."tamtam" AS "u_tamtam", "u"."email" AS "u_email", "u"."phone" AS "u_phone", "u"."avatarUrl" AS "u_avatarUrl", \n  (SELECT CAST( CASE WHEN "r"."updatedAt" IS NOT NULL  \n  THEN ((SELECT "r"."updatedAt") / 1000) ELSE (SELECT EXTRACT(EPOCH FROM TO_TIMESTAMP("r"."date", \'YYYY-MM-DD hh24:mi:ss.z\') at time zone \'utc+03:00\')) END AS int))\n AS "r_sort", \n                        (\n                        SELECT COUNT(c.id) FROM "comment" AS "c" \n                        WHERE ("c"."documentId" = "d"."id" AND\n                        "c"."isShared" = true\n                        )\n                        ) AS "d_commentsCount" FROM "report" "r" INNER JOIN "folder" "folder" ON  AND (folder.report = "r"."id")  LEFT JOIN "document" "d" ON "d"."reportId"="r"."id" AND ("d"."reportId" = "r"."id" AND "d"."isLast" = true)  LEFT JOIN "member" "u" ON "u"."id"="d"."responsibleId") "distinctAlias" ORDER BY "distinctAlias".r_sort DESC, "r_id" ASC LIMIT 15',
```
can you please resolve this ASAP ? :)

**Comment by cleivson** [comment]
Happened here too. I've fixed by moving the join condition to the where condition, but it's counter-intuitive and a little bit annoying

**Comment by biern** [comment]
I've experienced this issue even when chaining `where` queries, without using any kind of join, something along the lines of
```
.createQueryBuilder()
.select(...)
.from(...)
.where('bar IS NULL')
.orWhere('foo = :parameter', { parameter: 1 })
.orWhereInIds(['x', 'y', 'z'])
```
In generated query parameters had incorrect order:
`["x","y","z",1]` instead of `[1,"x","y","z"]`
so that it queried roughly for `foo = 'x' OR id in ('y', 'z', 1)` instead of expected `foo = 1 OR id in ('x', 'y', 'z', 1)`

**Comment by juuxstar** [comment]
Ran into this as well just now.

Seeing as how a query object can have any of the "construction" methods (ie. where, join, etc..) called in any order (which is useful as you can then pass the query object around to different parts of the code), queryBuilder needs to be smart enough to figure which parameter goes where and properly keep track of them.

**Comment by evilworm** [comment]
this is a major design issue, just faced the same issue... this should be a high-priority bug!

**Comment by dylantf** [comment]
I just encountered this as well. Here's my setup with a simple ManyToMany User - Group relationship.

```\d users_groups
                                       Table "public.users_groups"
   Column   |           Type           | Collation | Nullable |                 Default
------------+--------------------------+-----------+----------+------------------------------------------
 id         | integer                  |           | not null | nextval('users_groups_id_seq'::regclass)
 user_id    | integer                  |           | not null |
 group_id   | uuid                     |           | not null |
Indexes:
    "PK_4644edf515e3c0b88e988522588" PRIMARY KEY, btree (id)
    "users_groups_user_id_group_id_unique" UNIQUE CONSTRAINT, btree (user_id, group_id)
Foreign-key constraints:
    "users_groups_group_id_foreign" FOREIGN KEY (group_id) REFERENCES groups(id)
    "users_groups_user_id_foreign" FOREIGN KEY (user_id) REFERENCES users(id)
```

Classes:
```typescript
class User {
  @PrimaryGeneratedColumn()
  id: number;
  
  @ManyToMany(() => Group, group => group.users)
  @JoinTable({
    name: 'users_groups',
    joinColumn: { name: 'user_id', referencedColumnName: 'id' },
    inverseJoinColumn: { name: 'group_id', referencedColumnName: 'id' },
  })
  groups: Group[];
}

class Group {
  @PrimaryGeneratedColumn({ type: 'uuid' })
  id: string;
  
  @ManyToMany(() => User, user => user.groups)
  @JoinTable({
    name: 'users_groups',
    joinColumn: { name: 'group_id', referencedColumnName: 'id' },
    inverseJoinColumn: { name: 'user_id', referencedColumnName: 'id' },
  })
  users: User[];
}
```

And creating a new record:
```typescript
function addUserToGroup(userId: number, groupId: string): Promise<Group> {
  const user = await getRepository(User).find(userId);
  const group = await getRepository(Group)
    .createQueryBuilder('group')
    .where('group.id = :groupId', { groupId })
    .leftJoinAndSelect('group.users', 'user')
    .getOne();

  group.users.push(user);

  return getRepository(Group).save(group);
}
```

This creates the following SQL:
```sql
INSERT INTO "users_groups"("group_id", "user_id") VALUES ($1, $2) -- PARAMETERS: [3,"467ce297-c355-4e6a-a6d8-70148544db61"]
query failed: INSERT INTO "users_groups"("group_id", "user_id") VALUES ($1, $2) -- PARAMETERS: [3,"467ce297-c355-4e6a-a6d8-70148544db61"]
error: error: invalid input syntax for type uuid: "3"
```

The parameters are bound in the wrong order. Strangely, doing the opposite query (so, `users.groups.push(group)` generates the correct SQL. I tried playing with the class definitions to make sure I wrote it correctly, but it doesn't even run if I try changing the `joinColumn`/`inverseJoinColumn`.

**Comment by juansb827** [comment]
still happening :(

**Comment by willyguevara** [comment]
Hello anyone have a workaround?

**Comment by juansb827** [comment]
> Hello anyone have a workaround?
I think putting inside a  `where` whatever filter you were going to put in the join should work

**Comment by willyguevara** [comment]
> > Hello anyone have a workaround?
> > I think putting inside a  `where` whatever filter you were going to put in the `innerJoin` should work

Putting in the`where` clause not working because the filtering is in the relation of a many to many and not in the entity itself. so it fails .

I ended up like workin ok like this:

```js
const fts = 'wil';
const groupIdList = "1,2,3";
const qb = repo.createQueryBuilder('user');
// qb.innerJoin('eltitle.folders', 'folder', 'folder.id = (:folderId)', { folderId }); // <-- this generates an issue with the orders in  parameters
qb.innerJoin('user.groups', 'group', 'group.id IN (:...groupIdList)');
qb.where('(user.firstname LIKE :firstNameFts OR user.lastname LIKE :lastNameFts)');
qb.setParameters({
    folderIdList,
    firstNameFts: `%${fts}%`,
    lastNameFts: `%${fts}%`,
});
```

**Comment by davidrapin** [comment]
I experience the same bug (backend: SQLite).
I confirm that the only workaround that works for me is to call `builder.setParameters({key: value})` with all query parameters just before running the query with all query parameters.

**Comment by sakibrahmanchy** [comment]
Faced this issue too with mysql, with like where 
```
      ...................
     qb.where('user.first_name LIKE :first_name', { firstName: '%abcd%'})
```

**Comment by willyguevara** [comment]
> Faced this issue too with mysql, with like where
> 
> ```
>       ...................
>      qb.where('user.first_name LIKE :first_name', { firstName: '%abcd%'})
> ```

i think that problem here is that `:first_name` needs to be `:firstName`

**Comment by sakibrahmanchy** [comment]
@willyguevara Well this is not the actual query, this is a dummy one. But the parameter values were set correctly, construction of parameters where wrong.

**Comment by filozof6** [comment]
Is the code base in such a bad state that this cannot be fixed when it is hanging here for 2 years already? :D

**Comment by imnotjames** [comment]
this should be fixed by #7821 

The issue had to do with the mixing of native parameters & traditional parameters.  We've effectively dropped the use of native parameters entirely & deprecated them.  Because of stuff like this.

**Comment by imnotjames** [comment]
> I've experienced this issue even when chaining `where` queries, without using any kind of join, something along the lines of
> 
> ```
> .createQueryBuilder()
> .select(...)
> .from(...)
> .where('bar IS NULL')
> .orWhere('foo = :parameter', { parameter: 1 })
> .orWhereInIds(['x', 'y', 'z'])
> ```
> 
> In generated query parameters had incorrect order:
> `["x","y","z",1]` instead of `[1,"x","y","z"]`
> so that it queried roughly for `foo = 'x' OR id in ('y', 'z', 1)` instead of expected `foo = 1 OR id in ('x', 'y', 'z')`

Tested with this example and it looks like it is indeed fixed - not sure when we are going to do a new release.

I really need to get a rolling "bleeding edge" build set up..

**Comment by pgagarinov** [comment]
@imnotjames 
I can confirm that this bug is not fixed.

If we have `@JoinTable` on both User and Group then the cascade `save` fails because TypeORM passes the parameters into `insert` in incorrect order:

```
query: START TRANSACTION
query: INSERT INTO "group"("id", "name") VALUES (DEFAULT, $1) RETURNING "id" -- PARAMETERS: ["group2"]
query: INSERT INTO "user"("first_name", "last_name", "age") VALUES ($1, $2, $3) RETURNING "id" -- PARAMETERS: ["Timber2","Saw2",25]
query: INSERT INTO "users_groups"("group_id", "user_id") VALUES ($1, $2) -- PARAMETERS: [1,"a6734e30-3f41-4920-b88f-5e23fae7d851"]
query failed: INSERT INTO "users_groups"("group_id", "user_id") VALUES ($1, $2) -- PARAMETERS: [1,"a6734e30-3f41-4920-b88f-5e23fae7d851"]
error: error: invalid input syntax for type uuid: "1"
    at Parser.parseErrorMessage (/home/peter/_Git/_AoU/typeorm-exp/node_modules/pg-protocol/src/parser.ts:369:69)
    at Parser.handlePacket (/home/peter/_Git/_AoU/typeorm-exp/node_modules/pg-protocol/src/parser.ts:188:21)
    at Parser.parse (/home/peter/_Git/_AoU/typeorm-exp/node_modules/pg-protocol/src/parser.ts:103:30)
    at Socket.<anonymous> (/home/peter/_Git/_AoU/typeorm-exp/node_modules/pg-protocol/src/index.ts:7:48)
    at Socket.emit (events.js:400:28)
```
The complete example is here: https://github.com/pgagarinov/typeorm-jointable-bug

Here is the gist of it:

```
@Entity()
export class User {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({name: "first_name"})
  firstName: string;

  @Column({name: "last_name"})
  lastName: string;

  @Column()
  age: number;
  @ManyToMany(() => Group, group => group.users)
  @JoinTable({
    name: 'users_groups',
    joinColumn: { name: 'user_id', referencedColumnName: 'id' },
    inverseJoinColumn: { name: 'group_id', referencedColumnName: 'id' },
  })
  groups: Group[];
}

import { User } from "./User"
import {PrimaryGeneratedColumn, JoinTable, ManyToMany, Entity, Column} from "typeorm";


@Entity()
export class Group {
  @PrimaryGeneratedColumn("uuid")
  id: string;

  @Column()
  name: string;
  
  @ManyToMany(() => User, (user: { groups: any; }) => user.groups, { cascade: true })
  @JoinTable({
    name: 'users_groups',
    joinColumn: { name: 'group_id', referencedColumnName: 'id' },
    inverseJoinColumn: { name: 'user_id', referencedColumnName: 'id' },
  })
  users: User[];
}

createConnection().then(async connection => {

    const user = new User();
    user.firstName = "Timber2";
    user.lastName = "Saw2";
    user.age = 25;

    const group = new Group();
    group.name = "group2";

    console.log("Trying something risky...");
    group.users = [];
    group.users.push(user);
    
    connection.manager.getRepository(Group).save(group);

}).catch(error => console.log(error));

```

**Comment by dylantf** [comment]
@imnotjames another confirmation that this is not fixed as of 0.2.40
I'm not 100% sure about this but it seems like something _might_ have been fixed since I upgraded from 0.2.36 to 0.2.40, I had one select query failing before and passing after.

What is still definitely broken for me is on deletes, the parameters are still bound in the wrong order. Removing the "reverse" mapping (the `foo => foo.bars` part in `@ManyToMany(() => Foo, foo => foo.bars)`) is needed to fix the wrong bind order.
